clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -845,42 +844,93 @@ raw2trace_t::process_header(raw2trace_thread_data_t *tdata)
 }
 
 #ifdef BUILD_PT_POST_PROCESSOR
+
+std::unique_ptr<pt_data_buf_t>
+raw2trace_t::get_next_kernel_entry(raw2trace_thread_data_t *tdata,
+                                   std::unique_ptr<pt_metadata_buf_t> &pt_metadata,
+                                   uint64_t expected_syscall_idx)
+{
+    DR_ASSERT(tdata->kthread_file != nullptr);
+    if (tdata->kthread_file->eof())
+        return nullptr;
+    if (!tdata->pt_metadata_processed) {
+        log(2, "Reading PT metadata for tid " INT64_FORMAT_STRING "\n", tdata->tid);
+        pt_metadata = std::unique_ptr<pt_metadata_buf_t>(new pt_metadata_buf_t());
+        if (!tdata->kthread_file->read((char *)&pt_metadata->header[0],
+                                       PT_METADATA_PDB_HEADER_SIZE)) {
+            tdata->error = "Unable to read the PDB header of PT metadata from kernel "
+                           "thread log file";
+            return nullptr;
+        }
+
+        if (!tdata->kthread_file->read((char *)&pt_metadata->metadata,
+                                       sizeof(pt_metadata->metadata))) {
+            tdata->error = "Unable to read the PT metadata from kernel thread log file";
+            return nullptr;
+        }
+    }
+    log(2,
+        "Reading PT data header for tid " INT64_FORMAT_STRING
+        " expected syscall idx " INT64_FORMAT_STRING "\n",
+        tdata->tid, expected_syscall_idx);
+    std::unique_ptr<pt_data_buf_t> pt_data(new pt_data_buf_t);
+    if (!tdata->kthread_file->read((char *)&pt_data->header[0],
+                                   PT_DATA_PDB_HEADER_SIZE)) {
+        if (tdata->kthread_file->eof()) {
+            VPRINT(1, "Finished decoding all PT data for thread %d\n", tdata->tid);
+            return nullptr;
+        }
+        tdata->error = "Unable to read the PDB header of next syscall's PT data "
+                       "from kernel thread log file";
+        return nullptr;
+    }
+    uint64_t pid = pt_data->header[PDB_HEADER_PID_IDX].pid.pid;
+    uint64_t tid = pt_data->header[PDB_HEADER_TID_IDX].tid.tid;
+    uint64_t syscall_idx = pt_data->header[PDB_HEADER_SYSCALL_IDX_IDX].syscall_idx.idx;
+    uint64_t sysnum = pt_data->header[PDB_HEADER_SYSNUM_IDX].sysnum.sysnum;
+    uint64_t syscall_args_num =
+        pt_data->header[PDB_HEADER_NUM_ARGS_IDX].syscall_args_num.args_num;
+    uint64_t pt_data_size =
+        pt_data->header[PDB_HEADER_DATA_BOUNDARY_IDX].pt_data_boundary.data_size -
+        PT_DATA_PDB_HEADER_SIZE - syscall_args_num * sizeof(uint64_t);
+
+    log(2,
+        "Reading PT data for syscall_idx " INT64_FORMAT_STRING
+        " size " INT64_FORMAT_STRING " tid " INT64_FORMAT_STRING
+        " pid " INT64_FORMAT_STRING " num " INT64_FORMAT_STRING "\n",
+        syscall_idx, pt_data_size, tid, pid, sysnum);
+    pt_data->data.reset(new uint8_t[pt_data_size]);
```

Comment:
Not sure I completely follow this: you allocate memory on line 876 and read the header.  You pull out pid, tid, etc. in local vars lines 887-895 -- but then you throw away the header and never use those locals (except `pt_data_size`).  So the header is only needed for the size?  Should those other locals be inside a debug-only construct or sthg, or only extracted in the log()?

Reply:
We do need syscall_num and syscall_args_num (when we want to reconstruct the memory address  trace), we need the syscall_idx to ensure we got the correct syscall next (I should change that from an assert to a release build error probably). The pid and tid can potentially be omitted.

How about if I do it in a follow up PR. I'm trying to make this a strictly refactor PR.

Reply:
Added a TODO for now.

Reply:
> I should change that from an assert to a release build error probably

Oh it's already that, not just a DR_ASSERT.